### ISSUE TITLE
"No installers" error when try to execute letsencrypt-auto from subprocess.Popen

### ISSUE THREAD

**Issue by akamensky** [certbot/certbot#2371]
I am trying to automate my certificates renewal (in short check NotAfter of certificate and if past certain threshold then call letsencrypt), but I get very weird behavior.

The error message is like this:

```
# python ssl-checker.py -c /etc/ssl-checker/ssl-checker.conf
Updating letsencrypt and virtual environment dependencies......
Requesting root privileges to run with virtualenv: /root/.local/share/letsencrypt/bin/letsencrypt certonly --webroot --renew-by-default --agree-tos --webroot-path /var/www/html/ -d www.deleted.com -d deleted.com
usage:
  letsencrypt-auto [SUBCOMMAND] [options] [-d domain] [-d domain] ...

The Let's Encrypt agent can obtain and install HTTPS/TLS/SSL certificates.  By
default, it will attempt to use a webserver both for obtaining and installing
the cert. Major SUBCOMMANDS are:

  (default) run        Obtain & install a cert in your current webserver
  certonly             Obtain cert, but do not install it (aka "auth")
  install              Install a previously obtained cert in a server
  revoke               Revoke a previously obtained certificate
  rollback             Rollback server configuration changes made during install
  config_changes       Show changes made to server config during installation
  plugins              Display information about installed plugins
letsencrypt: error: unrecognized arguments: certonly --webroot --renew-by-default --agree-tos --webroot-path /var/www/html/
```

So it tells me that `certonly` is not recognized as an argument, but when I just simply copy paste line from above (the one after `Requesting root privileges to run with virtualenv:`), it works fine and updates certificates in the destination location.

In my code I am using `child = subprocess.Popen(args, shell=True)` where `args` is a list of arguments such as `args = ['/root/letsencrypt/letsencrypt-auto', 'certonly', '--webroot', '--renew-by-default', '--agree-tos', '--webroot-path'...]`

**Comment by bmw** [comment]
Are you still having problems with this? I'm unable to recreate the exact problem, but I would suggest either not setting `shell=True` or using `args='/root/letsencrypt-auto certonly --webroot...'`.

If neither of these suggestions fix your problem, what OS are you experiencing this on?

**Comment by akamensky** [comment]
Yes. I do still have this problem. I tried different variations including passing all arguments as an entire string, disabling shell and so on. Always get same problem.

It sees the string correctly as you can see but does not take it as arguments for some reason.

**Comment by akamensky** [comment]
The issue happens on Debian 7 64 bit.

**Comment by akamensky** [comment]
When I execute exactly same command manually from terminal it works though

**Comment by quenenni** [comment]
I have the exact same problem (I think, if not, sorry, I'll create my own ticket)

What I understood is that letsencrypt doesn't accept parameters passed in a variable.
I'm doing it in a bash script:
This works:
`
letsencrypt-auto certonly --webroot --force-renewal --test-cert -m info@xxxx.net -t --agree-tos -w /home/owncloud/instances/xxxx/ -d cloud.xxxxx.net
`
This doesn't work:
`
param="--force-renewal --test-cert"; 
letsencrypt-auto certonly --webroot  $param -m info@xxxx.net -t --agree-tos -w /home/owncloud/instances/xxxx/ -d cloud.xxxxx.net
`
with the error:
`letsencrypt: error: unrecognized arguments: --force-renewal --test-cert`

Anything I'll pass via the variable won't be recognized by letsencrypt, like this for ex:
`
param=" certonly --webroot --force-renewal --test-cert -m info@xxxx.net -t --agree-tos -w /home/owncloud/instances/xxxx/ -d cloud.xxxxx.net"
letsencrypt-auto $param
`
`
letsencrypt: error: unrecognized arguments: certonly --webroot --force-renewal --test-cert -m info@xxxx.net -t --agree-tos -w /home/owncloud/instances/xxxx/ -d cloud.xxxxx.net
`

And if I copy / paste the command generated by my script in a terminal, it works.

-----
Debian wheezy
GNU bash, version 4.2.37(1)-release (x86_64-pc-linux-gnu)
Letsencrypt version: commit 01608d4d8d0a23578d3509799a597d0272513a40
Date:   Wed Mar 16 14:25:19 2016 -0700

**Comment by quenenni** [comment]
And for info, if my variable is empty, the script works:
`
param=""; 

letsencrypt-auto certonly --webroot $param -m info@xxxx.net -t --agree-tos -w /home/owncloud/instances/xxxx/ -d cloud.xxxxx.net
`

**Comment by nitmir** [comment]
@quenenni This is more a shell issue than a letsencrypt issue. For me using bash, it certainly work:
```
# params="-v renew"
# ./letsencrypt-auto $params
Checking for new version...
Requesting root privileges to run letsencrypt...
   /root/.local/share/letsencrypt/bin/letsencrypt -v renew
2016-03-17 19:31:46,513:INFO:letsencrypt.cli:Saving debug log to /var/log/letsencrypt/letsencrypt.log
Processing /etc/letsencrypt/renewal/munin.genua.fr.conf
2016-03-17 19:31:46,548:INFO:letsencrypt.cli:Cert not yet due for renewal
Processing /etc/letsencrypt/renewal/power.genua.fr.conf
2016-03-17 19:31:46,550:INFO:letsencrypt.cli:updating legacy http01_port value
2016-03-17 19:31:46,552:INFO:letsencrypt.cli:Cert not yet due for renewal

The following certs are not due for renewal yet:
  /etc/letsencrypt/live/munin.genua.fr/fullchain.pem (skipped)
  /etc/letsencrypt/live/power.genua.fr/fullchain.pem (skipped)
No renewals were attempted.
```

I do not get (note that to have the same kind of error as you, I have to put several quotes):
```
# ./letsencrypt-auto "\"$params\""
Checking for new version...
Requesting root privileges to run letsencrypt...
   /root/.local/share/letsencrypt/bin/letsencrypt "-v renew"
usage: 
  letsencrypt-auto [SUBCOMMAND] [options] [-d domain] [-d domain] ...

The Let's Encrypt agent can obtain and install HTTPS/TLS/SSL certificates.  By
default, it will attempt to use a webserver both for obtaining and installing
the cert. Major SUBCOMMANDS are:

  (default) run        Obtain & install a cert in your current webserver
  certonly             Obtain cert, but do not install it (aka "auth")
  install              Install a previously obtained cert in a server
  renew                Renew previously obtained certs that are near expiry
  revoke               Revoke a previously obtained certificate
  rollback             Rollback server configuration changes made during install
  config_changes       Show changes made to server config during installation
  plugins              Display information about installed plugins
letsencrypt: error: unrecognized arguments: "-v renew"

```
You can try to do `eval ./letsencrypt-auto $params` to force the variable resolution before it is passed to `letsencrypt-auto`…

**Comment by quenenni** [comment]
@nitmir

Thanks for your quick answer.

It works too with your code.

Could you try with the certonly subcommand?

**Comment by nitmir** [comment]
@quenenni  It works too:
```
root@power:/etc/letsencrypt/utils# param=" certonly --webroot --force-renewal --test-cert -m info@xxxx.net -t --agree-tos -w /home/www/letsencrypt/ -d cloud.xxxxx.net";
root@power:/etc/letsencrypt/utils# ./letsencrypt-auto $paramChecking for new version...
Requesting root privileges to run letsencrypt...
   /root/.local/share/letsencrypt/bin/letsencrypt certonly --webroot --force-renewal --test-cert -m info@xxxx.net -t --agree-tos -w /home/www/letsencrypt/ -d cloud.xxxxx.net
Failed authorization procedure. cloud.xxxxx.net (http-01): urn:acme:error:connection :: The server could not connect to the client to verify the domain :: DNS problem: NXDOMAIN looking up A for cloud.xxxxx.net

IMPORTANT NOTES:
 - If you lose your account credentials, you can recover through
   e-mails sent to info@xxxx.net.
 - The following errors were reported by the server:

   Domain: cloud.xxxxx.net
   Type:   connection
   Detail: DNS problem: NXDOMAIN looking up A for cloud.xxxxx.net

   To fix these errors, please make sure that your domain name was
   entered correctly and the DNS A record(s) for that domain
   contain(s) the right IP address. Additionally, please check that
   your computer has a publicly routable IP address and that no
   firewalls are preventing the server from communicating with the
   client. If you're using the webroot plugin, you should also verify
   that you are serving files from the webroot path you provided.
 - Your account credentials have been saved in your Let's Encrypt
   configuration directory at /etc/letsencrypt. You should make a
   secure backup of this folder now. This configuration directory will
   also contain certificates and private keys obtained by Let's
   Encrypt so making regular backups of this folder is ideal.
```

(of course `cloud.xxxxx.net` do not exists, so boulder return an error, but arguments on the command line seems to be correctly interpreted)

**Comment by quenenni** [comment]
YES.. Finally, I found the problem..

The hard to follow IFS was set to a space before calling letsencrypt. That's why it was so difficult to understand why the error occurs so erratically.

The strange thing is that even if I set the IFS var to a line break just before (IFS=$'\n'), I still had the problems.
The solution was to unset it (unset IFS) and now all kind of parameters in my variable are correctly recognized by letsencrypt.

Damned.. So many hours lost because of this.. so sad.

@nitmir Big thanks for your time and help.

I guess we should delete our comments to clean this post as this isn't related to @akamensky's problem (and it's always annoying to have users start another thread in your post), neither a let's encrypt problem.

I'll delete my posts this evening, if you agree to it.

**Comment by wangkesen** [comment]
kesen@ubuntu-1gb-sgp1-01:~$ sudo apt install letsencrypt
[sudo] password for kesen: 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Suggested packages:
  python-letsencrypt-apache python-letsencrypt-doc
The following NEW packages will be installed:
  letsencrypt
0 upgraded, 1 newly installed, 0 to remove and 1 not upgraded.
Need to get 0 B/10.9 kB of archives.
After this operation, 24.6 kB of additional disk space will be used.
Selecting previously unselected package letsencrypt.
(Reading database ... 64010 files and directories currently installed.)
Preparing to unpack .../letsencrypt_0.4.1-1_all.deb ...
Unpacking letsencrypt (0.4.1-1) ...
Processing triggers for man-db (2.7.5-1) ...
Setting up letsencrypt (0.4.1-1) ...
kesen@ubuntu-1gb-sgp1-01:~$ letsencrypt
An unexpected error occurred:
IOError: [Errno 13] Permission denied: '/var/log/letsencrypt/letsencrypt.log'
Please see the logfile 'letsencrypt.log' for more details.
kesen@ubuntu-1gb-sgp1-01:~$ exit
logout
root@ubuntu-1gb-sgp1-01:~# letsencrypt
No installers seem to be present and working on your system; fix that or try running letsencrypt with the "certonly" command

**Comment by sydneyli** [comment]
On the off-chance that you're still encountering this problem, I might suggest asking for help at the [Let's Encrypt Community Forums](https://community.letsencrypt.org/c/help), where there's a large audience of people who are familiar with the project. You'll get faster help there!

Closing-- if an actionable Certbot bug comes out of a forum discussion, feel free to submit a new issue.
